src/ComputeSharp.D2D1/Shaders/Interop/Effects/ResourceManagers/D2D1ResourceTextureManagerImpl.ID2D1ResourceTextureManagerInternal.cs

```diff
@@ -0,0 +1,202 @@
+                    }
+
+                    @this->expectedDimensions = expectedDimensions;
+                }
+
+                // If this is the first time the method is called, just store the context.
+                // Before doing so, get an ID2D1Multithread instance, to ensure thread safety.
+                // For additional info, see docs on ID2D1EffectFactoryExtensions.GetD2D1Multithread.
+                if (@this->d2D1EffectContext is null)
+                {
+                    using ComPtr<ID2D1Multithread> d2D1Multithread = default;
+
+                    int hresult = effectContext->GetD2D1Multithread(d2D1Multithread.GetAddressOf());
+
+                    // If an ID2D1Multithread object is available, we can safely store the context
+                    if (Windows.SUCCEEDED(hresult))
+                    {
+                        _ = effectContext->AddRef();
+
+                        @this->d2D1EffectContext = effectContext;
+                        @this->d2D1Multithread = d2D1Multithread.Detach();
+                    }
+
+                    return hresult;
+                }
+
+                // Otherwise, just do nothing and return S_FALSE. This allows an existing resource texture
+                // manager to be shared across effects. If the resource cannot be shared, it will just
+                // return an error when effect is actually set, with ID2D1DrawInfo::SetResourceTexture.
+                return S.S_FALSE;
+            }
+        }
+
+        /// <inheritdoc cref="ID2D1ResourceTextureManagerInternal.GetResourceTexture"/>
+        [UnmanagedCallersOnly]
+        public static int GetResourceTexture(D2D1ResourceTextureManagerImpl* @this, ID2D1ResourceTexture** resourceTexture)
+        {
+            @this = (D2D1ResourceTextureManagerImpl*)&((void**)@this)[-1];
+
+            if (*resourceTexture is not null)
+            {
+                return E.E_POINTER;
+            }
+
+            lock (@this->lockHandle.Target!)
+            {
+                // If the effect context is null, it means Initialize has not been called yet
+                if (@this->d2D1EffectContext is null)
+                {
+                    return E.E_NOT_VALID_STATE;
+                }
+
+                // If the texture has already been created, just return it
+                if (@this->d2D1ResourceTexture is not null)
+                {
+                    @this->d2D1ResourceTexture->AddRef();
+
+                    *resourceTexture = @this->d2D1ResourceTexture;
+
+                    return S.S_OK;
+                }
+
+                // If the data is null at this point, it means CreateResourceTexture has not been called yet
+                if (@this->data is null)
+                {
+                    return E.E_NOT_VALID_STATE;
+                }
+
+                // Create the resource now, as it hasn't been created yet
+                int result = @this->d2D1EffectContext->CreateResourceTexture(
```

Comment:
you need to enter the factory lock before calling into `ID2D1EffectContext`

Reply:
Not needed, see comment above